pep/002-mixin-versioning.md

```diff
@@ -0,0 +1,352 @@
+
+
+#### porter mixins upgrade
+
+When global is false, bump the locked version of the mixin per the version range defined in the porter.yaml, and then update the bundle mixin cache.
+This mode is useful for dependabot style CI integrations that automatically keep a bundle up-to-date.
+
+When global is true, upgrade the specified mixin to the latest version in the global bundle cache.
+
+#### porter mixins uninstall
+
+❓ I am not sure if it makes sense to have keep this command? At the bundle level, removing a mixin doesn't seem to have a use case. They could just stop using it. However at the global cache level, what would really be useful is to clear the cache, or removing infrequently used mixin versions. I think cache management like that should be out-of-scope.
+
+
+### Mixin Declaration
+
+Bundle authors can optionally specify the source and version of the mixin that should be used with the bundle. Different versions/sources of a mixin cannot be used within the same bundle. 
+When a source is not specified it defaults to the official porter mixin feed, and the version to latest (similar to how the `porter mixins install` command works).
+
+```yaml
+mixins:
+    - arm:
+        mixin:
+            version: v1.2.3 # accepts a semver range, e.g. 1.x, 1.2.x
+            source: https://example.com/feed.xml or https://github.com/getporter/terraform-mixins/downloads
+        clientVersion: v1.2.3
+    - terraform
+```
+
+❓ If possible, I want to avoid using complex semver ranges, e.g. ~1.2, ^1.1, >= 1.2 < 3.0.0 || >= 4.2.3, etc. The x notation is very easy to understand. Maybe we support whatever masterminds/semver supports but only document the x notation.
+
+The mixin schema command should return information about the acceptable schema for the mixin's configuration.
+Porter manages injecting the properties for `mixin -> version|source` when it assembles the final schema during `porter schema`.
+
+```json
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "definitions": {
+    "mixinConfig": {
+      "type": "object",
+      "properties": {
+        "clientVersion": {
+          "type": "string"
+        }
+      },
+      "additionalProperties": false
+    }    
+  },
+  "type": "object",
+  "properties": {
+    "install": {
+      "type": "array",
+      "items": {
+        "$ref": "#/definitions/installStep"
+      }
+    },
+    "mixins": {
+      "type": "array",
+      "items": {
+        "$ref": "#/definitions/mixinConfig"
+      }
+    }
+  }
+}
+```
+
+### Mixin Resolution
+
+A new command, `porter mixins download`, downloads the mixins used by the bundle the local bundle mixin cache directory.
+
+```
+$ porter mixins download
+```
+
+1. Resolve the sources and versions of the mixins used by the bundle.
+  If porter.lock.yaml exists, then any existing locked versions are used when the mixin name and source matches.
+1. Identify if any required mixins are not in the cache and then download them.
```

Comment:
Is the cache mentioned here the global or local bundle variant?  (If the latter, the statement below is slightly confusing.)

And/or, is this command intended to also have the `[--global=false]` flag, similar to others mentioned above?

Reply:
This is a bundle level command. I am not sure what the global cache management command would be that does something similar. I think if there is sufficient interest we could do a follow-up PEP for global cache management.